### ISSUE TITLE
Difficulties with inserting into a self many-to-many relation

### ISSUE THREAD

**Issue by KingZee** [Vincit/objection.js#1462]
I have an article table, where an article can cite multiple articles. Each article can also have a single author.

Article model : 

```js
class Article extends DefaultModel {
  static get tableName() {
    return "articles";
  }
  static get relationMappings() {
    return {
      author: {
        relation: DefaultModel.BelongsToOneRelation,
        modelClass: "Author",
        join: {
          from: "articles.authorId",
          to: "authors.id"
        }
      },
      mentions: {
        relation: DefaultModel.ManyToManyRelation,
        modelClass: "Article",
        join: {
          from: "articles.id",
          through: {
            from: "mentions.citingArticle",
            to: "mentions.citedArticle"
          },
          to: "articles.id"
        }
      }
    };
  }
}
```

Mention model : 
```js
class Mention extends DefaultModel {
  static get tableName() {
    return "mentions";
  }
  static get idColumn() {
    return ["citingArticle", "citedArticle"];
  }
}
```

What I'm trying to do, is insertGraph, the main article + the articles that are mentioned in it, here is what I have : 

```js
async function insertData(fullArticle) {
  const articleTx = await transaction(Article.knex(), async tx => {
    const references = fullArticle.references
      .map(articleObj => {
          return {
            ...articleObj.article,
            author: articleObj.author
          };
      });

    const article = await Article.query(tx).insertGraph({
      ...fullArticle.article,
      author: fullArticle.author,
      mentions: references[0]
    });

    return article;
  });

  console.log(articleTx);
}
```

I know this is only inserting the main row + the first row, but I ran into multiple problems : 

1. The mention model has a composite key, when I try to insert using the code above I get a `(node:5944) UnhandledPromiseRejectionWarning: error: column "citedArticle" of relation "mentions" does not exist`. 
2. I also got the same error when I try to add `"#id"` to the main article mentions relation, and `"#ref"` to the cited article,  and insert them in 2 different objects.
3. A third problem I ran into here, was a unique constraint on both the citing and the cited article authors. What if both of them were written by the same author? I have a unique string column in the authors table that isn't the id, when I tried to insert I got `duplicate key value violates unique constraint "author_isn"`. I'm not sure how I can update the row to reference the existing author id if that happens.

Thank you.

**Comment by koskimas** [comment]
Well, the error says it all. There's no 'citedArticle' column in the `mentions` table. You also need to define the join table model class in tje relation, which you aren't doing.

**Comment by KingZee** [comment]
Hi! Thanks for your response : 

There is definitely a citedArticle column in the mentions table. :) I feel like you're undermining the problem I'm having, I have went through the docs more than once, and tried several things before coming here. Here is the console after me doing these actions in succession : 

1. Insert a single article with the exact query displayed in my first message, except with the mentions commented out, which is a graph insert of 4 tables. (even though I removed the tables that weren't relevant in my first message) 

2. Manually insert a row in the mentions table in the database, and run a simple query through objection that displays its data, you can clearly see it's fetching it correctly, and i had to use the same id for both rows because it's the only article inserted so far. 

3. Run the query displayed above, with the error showing that it couldn't find the row.

![Code_2019-09-07_09-07-38](https://user-images.githubusercontent.com/8393986/64471967-942b5480-d14f-11e9-9f6e-e4ebde35f24d.jpg)

Also, I genuinely don't understand what you mean by "need to define the join table model class in the relation, which you aren't doing". In the docs here : https://vincit.github.io/objection.js/guide/relations.html#examples

in the manytomany example, I don't see anything mentioned about the "persons_movies" model? if it is declared as a model at all? From what I understand, I am trying to create a many to many relation between Article and Article, through Mention. which I hope is the correct syntax like I showed in the Article model above. Is there anything else I need to add in the Mention model?

**Comment by koskimas** [comment]
I don't know what to tell you. That error comes directly from the database, not from objection. Maybe you have a typo in the table/column name somewhere? Maybe you are using `knexSnakeCaseMappers` and forgot to camelCase some column. I can debug this for you if you provide a reproduction.

**Comment by koskimas** [comment]
The mention model is not used at all by the way. You need to define it in the `through` object as said in the documentation and in all the example projects. See the `RelationThrough` object's documentation

**Comment by koskimas** [comment]
https://vincit.github.io/objection.js/api/types/#type-relationthrough

**Comment by KingZee** [comment]
Hi, thanks a lot for your time. DefaultModel has snake case mapping which all other models inherit from, I tried adding a modelClass inside the through attribute but it still didn't work, I think I made a reproductible repo : https://github.com/KingZee/objection_test
I'm not sure if it's something I'm completely missing.
Thanks again.

**Comment by koskimas** [comment]
@KingZee So the reason really was that you really don't have the `citedArticle` column. You have a `cited_article` column instead.

**Comment by KingZee** [comment]
Thanks again, I'm a bit ashamed because I could swear I tried that. It works now.

**Comment by KingZee** [comment]
Hi @koskimas, sorry to mention you again, I was figuring out how to structure my graph upsert to reference some nested models,

```
    let testInput = {
      title: "test article",
      journal: {
        title: "journal 2",
        issn: "200"
      },
      mentions: [
        {
          title: "other article",
          journal: {
            "#id": "journal1",
            title: "journal 1",
            issn: "100",
            publisher: {
              "#id": "publisher1",
              name: "same name",
              doiPrefix: "10"
            }
          }
        },
        {
          title: "different article",
          journal: {
            "#ref": "journal1",
            title: "journal 2",
            issn: "100",
            publisher: {
              "#ref": "publisher1",
              name: "same name",
              doiPrefix: "10"
            }
          }
        }
      ]
    };
```

the top-level relations work great with multiple rows when i use "#id" "#ref", but i think i'm having a problem with the publisher which is nested inside a journal which belongs to an article. when i try to insert the example above i get a : `Error: no reference found`.

Also I'm using `const article = await Article.query(tx).upsertGraph(testInput, {
      relate: true,
      unrelate: true,
      insertMissing: true,
      noDelete: true
    });` to be able to insert new rows even without specifying an id.

**Comment by hirokihokari** [comment]
Just leaving here something I noticed myself.

I was having a similar issue where the error claims "columnName does not exists" when it indeed exists(so I thought).
It was mentioned

> you don't have a citedArticle column. You have a cited_article column instead.

and I thought "duh. and the `snakeCaseMappers ` option takes care of it". Indeed in my other insert queries with camelCase parameters succeeded. But then I realized that the debug message for my other queries showed already converted column name as in, `insert into table_name ("column_name") values ...`, not `insert into table_name ("columnName") values ...` which was the case with this query I was having issue with.

Turns out, looks like queries using `relatedQuery()` or `$relatedQuery()` won't use this `snakeCaseMappers` option or objection(or the objection objection or whatever) and I needed `knexSnakeCaseMapper` option for knex to be specified, which I did not, since the page states `There are two ways to achieve this:`, which I interpreted as "I only need one of these".
